src/git.ts

```diff
@@ -0,0 +1,259 @@
+      for (const repo of repos) {
+        names.push(repo.full_name);
+        ownerRepoMap.set(repo.name.toLowerCase(), repo);
+      }
+      ++page;
+    } while (repos.length > 0);
+
+    // TODO(usergenic): Update this function to support user repos as well as
+    // the org repos.
+    return names;
+  }
+
+  /**
+   * Given a nodegit repository, issue a git pull to bring it up to date.
+   */
+  async update(nodegitRepo: nodegit.Repository) {
+    await nodegitRepo.fetch('origin', this._cloneOptions.fetchOpts);
+  }
+
+  /**
+   * @returns the current user info from GitHub.  The current user is
+   * determined by the token being used.
+   */
+  async getCurrentUser(): Promise<GitHub.User> {
+    return promisify(this._github.user.get)({});
+  }
+}
+
+/**
+ * Checks out a branch with a given name on a repo.
+ *
+ * @returns the nodegit Branch object for the new branch.
+ */
+export async function checkout(
+    nodegitRepo: nodegit.Repository,
+    checkoutRef?: string): Promise<nodegit.Repository> {
+  const cwd = nodegitRepo.workdir();
+  const ref = typeof checkoutRef === 'string' ? checkoutRef : 'master';
+  return new Promise<nodegit.Repository>(
+      (resolve, reject) => (child_process.exec(
+          `git checkout ${ref}`, {cwd: cwd}, (error, stdout, stderr) => {
+            if (error) {
+              console.log(ref);
+              console.log(`Error checking out ${ref} in : ${cwd}`);
+            }
+            resolve(nodegitRepo);
+          })));
+}
+
+/**
+ * @returns a string representation of a RepoRef of the form:
+ *     "name:org/repo#ref"
+ */
+export function serializeGitHubRepoRef(repoRef: GitHubRepoRef): string {
+  const checkoutRef = repoRef.checkoutRef ? `#${repoRef.checkoutRef}` : '';
+  return `${repoRef.ownerName}/${repoRef.repoName}${checkoutRef}`;
+}
+
+/**
+ * @returns a GitHubRepoRef resulting from the parsed string of the form:
+ *     `ownerName/repoName[#checkoutRef]`
+ */
+export function parseGitHubRepoRefString(refString: string): GitHubRepoRef {
+  const hashSplit = refString.split('#');
+  const slashSplit = hashSplit[0].split('/');
+
+  if (slashSplit.length !== 2 || hashSplit.length > 2) {
+    throw(`Repo '${refString}' is not in form user/repo or user/repo#ref`);
+  }
+
+  const owner = slashSplit[0];
+  const repo = slashSplit[1];
+  const ref = hashSplit[1];
+
+  return {ownerName: owner, repoName: repo, checkoutRef: ref};
+}
+
+/**
+ * @returns whether the matcherRef matches the targetRef, which allows for the
+ *     case-insensitive match as well as wildcards.
+ */
+export function matchRepoRef(
+    matcherRef: GitHubRepoRef, targetRef: GitHubRepoRef): boolean {
+  return util.wildcardRegExp(matcherRef.ownerName).test(targetRef.ownerName) &&
+      util.wildcardRegExp(matcherRef.repoName).test(targetRef.repoName);
+}
+
+/**
+ * @returns true if the repo is actually a response redirecting to another repo
+ */
+function isRedirect(repo: GitHub.Repo): boolean {
```

Comment:
The next version of this node package supports followRedirects behavior so this code will hopefully go away after upgrading to it (I tried a quick upgrade but had some issues so I'll have to hold off after this PR.)